### DISCUSSION TITLE
`eleventy clean` to remove output (👍 128)

### DISCUSSION THREAD

**Discussion by zachleat** [11ty/eleventy#3967] [Enhancement Queue]
See https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/remove.md

* **Must** not delete any files not generated by the tool.
* Should probably implement as a `destroy` method on `Template`

Cheating methods that probably aren’t good enough:
* Delete output directory if input and output don’t match (or if input isn’t a directory of output)

**Comment by zachleat** [comment]
This repository is now using lodash style issue management for enhancements (see https://twitter.com/samselikoff/status/991395669016436736)

This means enhancement issues will now be closed instead of leaving them open. The enhancement backlog can be found here: https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+

**Comment by veleek** [comment]
Just as a clarification, if I remove an input file eleventy will no longer be able to tell that it was the one that generated a given output file.  Should we expect that `eleventy clean` will not remove this file, or would we need some way to track that info?

Could we make an argument that for most people to use something like `git clean` on the eleventy output folder is a better option.  Presumably in most cases, any files that people care about in the output directory are going to be under source control so that's the system that has final word on what should be there.  `git clean -e node_modules` covers all the bases for me including stuff that's generated by gulp and other tasks.  I image that most of the other upvoters have useful reasons for wanting this.  It would be interesting to see what sort of specific scenarios they have that depends on this.

**Comment by tomglynch** [comment]
Currently this is not implemented, so the workaround is to just delete the `/_site` folder and re-run `npx @11ty/eleventy`. This works fine as long as you have not made changes in files within `/_site`.

**Comment by patricknelson** [comment]
Not super cross OS compatible (but I'm on Windows, I just run Bash on Cygwin), but I just use the command `rm -rf ./public/ && eleventy` and similar. But yeah ideally if you're working in a team with mixed host OS's, you'd have a simpler command you could run that'd "just work" everywhere. 🤔

**Comment by RichardJECooke** [comment]
I'm confused. If all enhancements are marked as closed how do we know if this is done or not?

**Comment by RichardJECooke** [comment]
Note, this command should also be available when running `--serve`.  Something like `npx @11ty/eleventy --serve --clean` so that deleted pages are removed from output as you work and review in the browser.

**Comment by veleek** [comment]
> I'm confused. If all enhancements are marked as closed how do we know if this is done or not?

See the referenced thread: https://twitter.com/samselikoff/status/991395669016436736

If it's tagged with Feature and Votes Needed it hasn't been completed.

**Comment by kentaroi** [comment]
It's just an experiment, but I implemented the feature and published it as a plugin, which deletes files in the output directory, previously created by Eleventy but no longer created/updated, at the end of each build.
https://github.com/kentaroi/eleventy-plugin-clean
